### DISCUSSION TITLE
How can I run queries inside an isolate

### DISCUSSION THREAD

**Discussion by SiroDevs** [simolus3/drift#2629] [Q&A]
I am using Dependency Injection in my app and I was working to work with Flutter Isolate to run my queries. However I keep getting errors of 'object is unsendable' when I try my isolate. I can't seem to get the best to do this.

I am working with the following classes: DbRepository >> ItemDao >> AppDatabase

DbRepository class
 
    import 'package:injectable/injectable.dart';

    import '../db/dao/item_dao.dart';
    import '../model/base/item.dart';

    @lazySingleton
    abstract class DbRepository {
      @factoryMethod
      factory DbRepository(ItemDao itemDao) = DbRepo;

      Future<void> saveItem(Item item);
      Future<void> editSong(Item song);
    }

    class DbRepo implements DbRepository {
      final ItemDao itemDao;

      DbRepo(this.itemDao);

      @override
      Future<void> saveItem(Item item) async => await itemDao.createItem(item);

      @override
      Future<void> editItem(Item item) async => await itemDao.updateSong(item);
    }

ItemDao class

    import 'package:drift/drift.dart';
    import 'package:injectable/injectable.dart';
    
    import '../../model/base/item.dart';
    import '../../model/tables/items_table.dart';
    import '../app_database.dart';
    
    part 'item_dao.g.dart';
    
    @lazySingleton
    abstract class ItemDao {
      @factoryMethod
      factory ItemDao(AppDatabase db) = _ItemDao;
    
      Future<int> createItem(Item item);
      Future<void> updateItem(Item item);
    }
    
    @DriftAccessor(tables: [ItemsTable])
    class _ItemDao extends DatabaseAccessor<AppDatabase>
        with _$_ItemDaoMixin
        implements ItemDao {
      _ItemDao(AppDatabase db) : super(db);
    
      @override
      Future<int> createItem(Item item) async {
        final sqlStatement = ItemsTableCompanion.insert(
          title: Value(item.title ?? ''),
          created: Value(item.created ?? ''),
        );
        result = await into(db.itemsTable).insert(sqlStatement);
      }
    
      @override
      Future<int> updateItem(Item item) async {
        final sqlStatement = ItemsTableCompanion(
          title: Value(item.title!),
          updated: Value(item.updated!),
        );
        await (update(db.itemsTable)..where((row) => row.id.equals(item.id)))
            .write(sqlStatement);
      }
    }

AppDatabase class
    
    import 'package:drift/drift.dart';
    
    import '../model/tables/items_table.dart';
    import 'dao/item_dao.dart';
    
    part 'app_database.g.dart';
    
    @DriftDatabase(
      tables: [ItemsTable],
      daos: [ItemDao],
    )
    class AppDatabase extends _$AppDatabase {
      AppDatabase(QueryExecutor db) : super(db);
    
      AppDatabase.connect(DatabaseConnection connection)
          : super.connect(connection);
    
      @override
      int get schemaVersion => 2;
    }

My Connection to the Database

    import 'dart:io';
    import 'dart:isolate';
    
    import 'package:drift/drift.dart';
    import 'package:drift/isolate.dart';
    import 'package:drift/native.dart';
    import 'package:path/path.dart';
    import 'package:path_provider/path_provider.dart';
    
    Future<DatabaseConnection> createDriftDatabaseConnection(String name) async {
      Directory dbFolder = await getApplicationDocumentsDirectory();
    
      final file = File(join(dbFolder.path, '$name.sqlite'));
    
      final receivePort = ReceivePort();
    
      await Isolate.spawn(
        _startBackground,
        _IsolateStartRequest(receivePort.sendPort, file.path),
      );
      final isolate = await receivePort.first as DriftIsolate;
      return isolate.connect();
    }
    
    void _startBackground(_IsolateStartRequest request) {
      final executor = NativeDatabase(File(request.targetPath));
      final driftIsolate = DriftIsolate.inCurrent(
        () => DatabaseConnection.fromExecutor(executor),
      );
      request.sendDriftIsolate.send(driftIsolate);
    }
    
    class _IsolateStartRequest {
      final SendPort sendDriftIsolate;
      final String targetPath;
    
      _IsolateStartRequest(this.sendDriftIsolate, this.targetPath);
    }

Now when running my isolate I use the following code which works for all other objects except for any related with the database

  static void fetchItems() async {
      final resultPort = ReceivePort();
      DbRepository db = getIt.get<DbRepository>();
      WebRepository api = getIt.get<WebRepository>();
  
      await Isolate.spawn(
        fetchAndSaveItems,
        [resultPort.sendPort, api, db], //This is where the code fails because db object is unsendable
      );
    }
  
    static Future<void> fetchAndSaveItems(List<dynamic> args) async {
      SendPort resultPort = args[0];
      WebRepository api = args[1];
      DbRepository db = args[2]; //fails
  
      List<Item>? items = [];
      var response = await api.fetchItems();
      var resp = jsonDecode(response.body);
      if (response.statusCode == 200) {
        List<dynamic> dataList = resp['data'];
        items = dataList.map((item) => Item.fromJson(item)).toList();
        if (items.isNotEmpty) {
          for (int i = 0; i < items.length; i++) {
            db.saveItem(items[i]); //I would be happy if I get this line to work too
          }
        }
      } 
      Isolate.exit(resultPort, items);
    }

alternatively when I don't try to pass my db object to the isolate everything works as below

     static void fetchItems() async {
      final resultPort = ReceivePort();
      DbRepository db = getIt.get<DbRepository>();
      WebRepository api = getIt.get<WebRepository>();
  
      await Isolate.spawn(
        fetchAndSaveItems,
        [resultPort.sendPort, api],
      );
      List<Item> items = await (resultPort.first) as List<Item>;
      if (items.isNotEmpty) {
        for (int i = 0; i < items.length; i++) {
          db.Item(items[i]);
        }
      }
    }
  
    static Future<void> fetchAndSaveItems(List<dynamic> args) async {
      SendPort resultPort = args[0];
      WebRepository api = args[1];
      DbRepository db = args[2];
  
      List<Item>? items = [];
      var response = await api.fetchItems();
      var resp = jsonDecode(response.body);
      if (response.statusCode == 200) {
        List<dynamic> dataList = resp['data'];
        items = dataList.map((item) => Item.fromJson(item)).toList();
      } 
      Isolate.exit(resultPort, items);
    }

I have tried to get an instance of `AppDatabase` as well as that of `DatabaseConnection` but still none of the objects can be sendable to the isolate. Any ideas to do it better than this like using the `DriftIsolate` would be helpful to me.

**Comment by frankvollebregt** [comment]
Have you taken a look at the documentation page about using isolates: https://drift.simonbinder.eu/docs/advanced-features/isolates/ in a simple case, it's quite straightforward and doesn't require much manual setup. Though if you want/need you can still do things manually, by sending the `SendPort`s over, but most of it is explained in the linked page I think

**Reply by simolus3** [reply]
I agree with this answer. Instances of your database class are stateful objects that can't be shared across isolates. Thanks to the `package:drift/isolate.dart` library though, you can share the underlying database connection across isolates and open another instance of your database class on the remote isolate. Drift tries it's best to make the connection sharing fully transparent - the two isolates have a shared view over the database and stream queries are synchronized as well.

You can use the [simple sharing](https://drift.simonbinder.eu/docs/advanced-features/isolates/#simple-sharing) feature for this, which is explained on the documentation site about isolates.
